utils/skopeo.py

```diff
@@ -0,0 +1,239 @@
+                          E.g. inspect, copy, ...
+        :type subcomand: str
+        :param *args: Additional positional arguments according
+                      to the sub-command.
+        :type *args: str
+        :param src_creds: (optional) The source repository
+                           credentials in the format
+                           "username:password".
+        :type src_creds: str
+        :param dest_creds: (optional) The destination repository
+                           credentials in the format
+                           "username:password".
+        :type dest_creds: str
+        """
+        cmd = [self.skopeo_cmd, subcomand]
+
+        if src_creds is not None:
+            cmd.append(f'--src-creds={src_creds}')
+        if dest_creds is not None:
+            cmd.append(f'--dest-creds={dest_creds}')
+        cmd.extend(args)
+
+        # Generating a safe cmd (no password exposed) for logging
+        safe_cmd = []
+        for item in cmd:
+            if 'creds' in item:
+                # Obfuscating the password with asterisks
+                safe_cmd.append(':'.join([item.split(':')[0], '******']))
+            else:
+                safe_cmd.append(item)
+        self.log.info('Running "%s"', ' '.join(safe_cmd))
+
+        if self.dry_run:
+            self.log.info(' DRY-RUN: Nothing really executed.')
+            return ''
+
+        result = subprocess.run(cmd, capture_output=True)
+
+        for line in result.stdout.decode().splitlines():
+            self.log.debug(' %s', line)
+
+        if result.returncode:
+            for line in result.stderr.decode().splitlines():
+                self.log.error(' %s', line)
+            raise SkopeoCmdError(f'exit code: {result.returncode}')
+
+        return result.stdout
+
+    def _parse_image_url(self, image):
+        """
+        Parser to split the image urls in its multiple components.
+
+        Images are provided as URLs. E.g.:
+            - docker.io/fedora
+            - docker.io/fedora:31
+            - docker://docker.io/user/fedora
+            - docker://registry.example.com:5000/repo01/centos:latest
+
+        Regardless the components provided in the URL, we have to make
+        sure that we can properly split each of them and, for those
+        not provided, assume safe defaults.
+
+        Example:
+            Considering the image URL "quay.io/app-sre/qontract-reconcile"
+
+        The data structure returned will be:
+            {'scheme': 'docker://',
+             'host': 'quay.io',
+             'image': 'app-sre/qontract-reconcile',
+             'tag': 'latest'}
+
+        :param image: The image url to be parsed.
+        :type image: str
+        :return: A data structure with all the parsed components of
+                 the image URL, already filled with the defaults for
+                 those not provided.
+        :rtype: dict
+        """
+        image_url_struct = re.search(
```

Comment:
would it be possible to leverage https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse here?

Reply:
There are some issues with using `urlparse` for this use case that I would have to work around:
- It requires the scheme to be provided, so I'd have to pre-process the url:
```
>>> urlparse('registry.example.com/apahim/fedora')
ParseResult(scheme='', netloc='', path='registry.example.com/apahim/fedora', params='', query='', fragment='')
```
- It does not parse the tag, when it exists, requiring some post-processing:
```
>>> urlparse('socker://registry.example.com/apahim/fedora:31')
ParseResult(scheme='socker', netloc='registry.example.com', path='/apahim/fedora:31', params='', query='', fragment='')
```
- On top of that, after parsed, you are not able to overwrite the attributes, so you have to create another data structure out of it so you can manipulate the values.

With all that, I think the regex gives us a very straight forward implementation.

Reply:
:+1: thanks for looking into it